server/app/services/geo/esri/EsriClient.java

```diff
@@ -0,0 +1,179 @@
+  }
+
+  /** Retries failed requests up to the provided value */
+  private CompletionStage<WSResponse> tryRequest(WSRequest request, MutableInt retries) {
+    CompletionStage<WSResponse> responsePromise = request.get();
+    responsePromise.handle(
+        (result, error) -> {
+          if (error != null || result.getStatus() != 200) {
+            if (retries.compareTo(new MutableInt(0)) > 0) {
+              retries.decrement();
+              return tryRequest(request, retries);
+            } else {
+              return responsePromise;
+            }
+          } else {
+            return CompletableFuture.completedFuture(result);
+          }
+        });
+
+    return responsePromise;
+  }
+
+  /**
+   * Returns address candidates from Esri's findAddressCandidates service
+   *
+   * <p>@see <a
+   * href="https://gisdata.seattle.gov/server/sdk/rest/index.html#/Find_Address_Candidates/02ss00000015000000/">Find
+   * Address Candidates</a>
+   */
+  public CompletionStage<Optional<JsonNode>> fetchAddressSuggestions(ObjectNode addressJson) {
+    if (this.ESRI_FIND_ADDRESS_CANDIDATES_URL.isEmpty()) {
+      return CompletableFuture.completedFuture(Optional.empty());
+    }
+    WSRequest request = ws.url(this.ESRI_FIND_ADDRESS_CANDIDATES_URL.get());
+    request.setContentType(ESRI_CONTENT_TYPE);
+    request.addQueryParameter("outFields", ESRI_FIND_ADDRESS_CANDIDATES_OUT_FIELDS);
+    // "f" stands for "format", options are json and pjson (PrettyJson)
+    request.addQueryParameter("f", ESRI_FIND_ADDRESS_CANDIDATES_FORMAT);
+    String address = addressJson.findPath(AddressField.STREET.getValue()).textValue();
+    System.out.println(address);
+    String address2 = addressJson.findPath(AddressField.LINE2.getValue()).textValue();
+    String city = addressJson.findPath(AddressField.CITY.getValue()).textValue();
+    String region = addressJson.findPath(AddressField.STATE.getValue()).textValue();
+    String postal = addressJson.findPath(AddressField.ZIP.getValue()).textValue();
+    if (address != null) {
+      request.addQueryParameter("address", address);
+    }
+    if (address2 != null) {
+      request.addQueryParameter("address2", address2);
+    }
+    if (city != null) {
+      request.addQueryParameter("city", city);
+    }
+    if (region != null) {
+      request.addQueryParameter("region", region);
+    }
+    if (postal != null) {
+      request.addQueryParameter("postal", postal);
+    }
+    MutableInt tries = new MutableInt(3);
+    return tryRequest(request, tries)
+        .thenApply(
+            res -> {
+              // return empty if still failing after retries
+              if (res.getStatus() != 200) {
+                return Optional.empty();
+              }
+              return Optional.of(res.getBody(json()));
+            });
+  }
+
+  /**
+   * Returns an {@link AddressSuggestionGroup} future and is the primary way CiviForm services
+   * should interact with the Esri API
+   */
```

Comment:
suggestion (blocking): state what the method does (calls the external ESRI service to retrieve address correction suggestions for the given address) as the first line of the javadoc since that what shows up in the method preview in most IDEs and code navigation tools. 

Document failure mode of returning an empty optional and what that indicates.

mention retry behavior

Reply:
Updated javadoc per suggestion. This method technically calls `fetchAddressSuggestions`, which calls the Esri service. Javadocs now reflect that.

Reply:
Since `fetchAddressSuggestions` is a private method, mentioning it in the javadoc isn't appropriate. Prefer to describe the behavior of the method directly.

Reply:
Updated and now reads:

```
/**
   * Calls an external Esri service to get address suggestions given the provided {@link Address}.
   * Takes the returned address suggestions to build an {@link AddressSuggestionGroup}.
   *
   * @return an optional {@link AddressSuggestionGroup} if successful, or an empty optional if the request fails.
   */
```